gusto/state.py

```diff
@@ -52,6 +53,109 @@ def __iter__(self):
+                group.createDimension("points", npts)
+                group.createDimension("geometric_dimension", dim)
+                var = group.createVariable("points", points.dtype,
+                                           ("points", "geometric_dimension"))
+                var[:] = points
+                group.createVariable(field_name,
+                                     field_creator(field_name).dat.dtype,
+                                     ("time", "points"))
+
+    def dump(self, field_creator, t):
+        """Evaluate and dump field data at points.
+
+        :arg field_creator: :class:`FieldCreator` for accessing
+            fields.
+        :arg t: Simulation time at which dump occurs.
+        """
+        with Dataset(self.filename, "a") as dataset:
+            # Add new time index
+            idx = dataset.dimensions["time"].size
+            dataset.variables["time"][idx:idx + 1] = t
+            for field_name, points in self.field_points:
+                vals = np.asarray(field_creator(field_name).at(points))
+                group = dataset.groups[field_name]
+                var = group.variables[field_name]
+                var[idx, :] = vals
+
+
+class DiagnosticsOutput(object):
+    def __init__(self, filename, diagnostics, description, create=True):
+        """Create a dump file that stores diagnostics.
+
+        :arg filename: The filename.
+        :arg diagnostics: The :class:`Diagnostics` object.
+        :arg description: A description.
+        :kwarg create: If False, assume that filename already exists
+        """
+        self.filename = filename
+        self.diagnostics = diagnostics
+        if not create:
+            return
+        with Dataset(filename, "w") as dataset:
+            dataset.description = "Diagnostics data for simulation {desc}".format(desc=description)
+            dataset.history = "Created {t}".format(t=time.ctime())
+            dataset.source = "Output from Gusto model"
+            dataset.createDimension("time", None)
+            var = dataset.createVariable("time", np.float64, ("time", ))
+            var.units = "seconds"
+            for name in diagnostics.fields:
+                group = dataset.createGroup(name)
+                for diagnostic in diagnostics.available_diagnostics:
+                    group.createVariable(diagnostic, np.float64, ("time", ))
+
+    def dump(self, diagnostic_fields, state, t):
+        """Dump diagnostics.
+
+        :arg diagnostic_fields: Iterable of :class:`DiagnosticField` objects.
+        :arg state: The :class:`State` at which to compute the diagnostic.
+        :arg t: The current time.
+        """
+        mapping = dict((df.name, df) for df in diagnostic_fields)
+        with Dataset(self.filename, "a") as dataset:
+            idx = dataset.dimension["time"].size
+            dataset.variables["time"][idx:idx + 1] = t
+            for name in self.diagnostics.fields:
+                field = mapping[name](state)
```

Comment:
OK, so now we ask for a `DiagnosticField` and immediately evaluate it to get the resulting `Function`.  Previously we evaluated the diagnostic fields up front, and relied on them magically being up to date in `State` when actually computing the diagnostics.